plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/ResourceCache.java

```diff
@@ -0,0 +1,177 @@
+ * Copyright (c) 2016 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v1.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package org.jboss.tools.openshift.internal.ui.models;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.eclipse.osgi.util.NLS;
+import org.jboss.tools.openshift.internal.core.Trace;
+
+import com.openshift.restclient.model.IProject;
+import com.openshift.restclient.model.IResource;
+
+public class ResourceCache {
+	private Map<String, Map<String, IResource>> cache = new HashMap<String, Map<String, IResource>>();
+
+	public void dispose() {
+		flush();
+	}
+
+	public void flush() {
+		synchronized (cache) {
+			cache.clear();
+		}
+	}
+
+	public void flush(String namespace) {
+		synchronized (cache) {
+			cache.remove(namespace);
+		}
+	}
+
+	/**
+	 * Key for caching an object
+	 * 
+	 * @param resource
+	 * @return
+	 */
+	private String getCacheKey(IResource resource) {
+		return getCacheKey(resource.getKind(), resource.getName());
+	}
+
+	private String getCacheKey(String kind, String name) {
+		return NLS.bind("{0}/{1}", new Object[] { kind, name });
+	}
+
+	@SuppressWarnings("unchecked")
+	public <T extends IResource> T getResource(String namespace, String kind, String name) {
+		synchronized (cache) {
+			Map<String, IResource> projectResources = cache.get(namespace);
+			if (projectResources != null) {
+				return (T) projectResources.get(getCacheKey(kind, name));
+			}
+			return null;
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	public <T extends IResource> Collection<T> getResources(String namespace, String kind) {
+		synchronized (cache) {
+
+			Map<String, IResource> projectResources = cache.get(namespace);
+			if (projectResources != null) {
+				return projectResources.values().stream().filter(r -> kind.equals(r.getKind())).map(r -> (T) r)
+						.collect(Collectors.toList());
+			}
+			return Collections.emptyList();
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	public <T extends IResource> Collection<T> getResources(String namespace) {
+		synchronized (cache) {
+			Map<String, IResource> projectResources = cache.get(namespace);
+			if (projectResources != null) {
+				return new ArrayList<T>((Collection<T>) projectResources.values());
+			}
+			return Collections.emptyList();
+		}
+	}
+
+	/**
+	 * 
+	 * @param resource
+	 * @return true if cached; false otherwise
+	 */
+	public boolean add(IResource resource) {
+		if (resource == null)
+			return false;
+		synchronized (cache) {
+			if (getCachedVersion(resource) != null) {
```

Comment:
Can cached version happen to be another obsolete instance getCachedVersion(resource) != resource?
Or that is prevented with duly work of remove and update methods?